src/ui/Components/CustomBoundary.tsx

```diff
@@ -0,0 +1,75 @@
+import React from "react";
+import Typography from "@mui/material/Typography";
+
+interface PlayerBoundaryProps {
+  children: React.ReactNode;
+}
+interface PlayerBoundaryState {
+  error?: Error;
+}
+/** Error boundary for custom content printed by the player using printRaw-like functions.
+ * Error boundaries are required to be class components due to no FC equivalent to componentDidCatch. */
+export class CustomBoundary extends React.Component<PlayerBoundaryProps, PlayerBoundaryState> {
+  state: PlayerBoundaryState;
+  constructor(props: PlayerBoundaryProps) {
+    super(props);
+    const error = probeChildrenForFunction([], props.children);
+    this.state = { error };
+  }
+  componentDidCatch(error: Error): void {
+    this.setState({ error });
+    console.warn("Error in custom react content:");
+    console.error(error);
+  }
+  render(): React.ReactNode {
+    if (this.state.error) {
+      // Typography is used because there are no default page styles.
+      // Span is used because it does not conflict with the DOM validation nesting (default Typography element of p is invalid at this location in dom tree)
+      return <Typography component={"span"}>Error in custom react content. See console for details.</Typography>;
+    }
+    return <Typography component={"span"}>{this.props.children}</Typography>;
+  }
+}
+
+/** Attempting to render a react tree containing a function as a child will cause an error that doesn't get displayed
```

Comment:
Doesn't get displayed correctly how?
This seems like rather heavyweight error-checking to be doing on every call, so it would be really nice to not need to do it.

Reply:
Yeah so the Boundary is there to try and catch any errors, that way they can be communicated to the player by displaying their content as an error message and directing them to the console so they are aware of the error they have in their code.

Without this checking, if the player provides a function, the game will display an empty span (which is usually invisible) instead of a span telling the player they have an error, and will throw a different error that the player would just have to happen to check the console to ever see, that only displayed once instead of every time the player attempts to render invalid content:

```js
const el = React.createElement;

function TestComponent(props) {
  return props.children;
}

/** @param {NS} ns */
export async function main(ns) {
  ns.tail();
  ns.tprint("Test");
  console.log("Test");
  ns.iKnowWhatImDoing();
  tprintRaw(TestComponent)
  ns.printRaw(
    el(TestComponent, null,
      el("span", null,
        el("div", null,
          el(TestComponent, null,
            TestComponent // TestComponent incorrectly used as a child directly
          )
        )
      )
    )
  );
  console.log("Test2")
  ns.tprint("Test2");

}
```
![image](https://github.com/bitburner-official/bitburner-src/assets/84951833/85046987-26bc-4c8e-a673-3f611cadd181)
![image](https://github.com/bitburner-official/bitburner-src/assets/84951833/666d0cde-aad6-493a-bd0a-815647a6a790)

Console on first run:
![image](https://github.com/bitburner-official/bitburner-src/assets/84951833/fbd07643-4caa-4c95-9660-2b7aa62d9375)

Console on subsequent runs (no error message at all):
![image](https://github.com/bitburner-official/bitburner-src/assets/84951833/dc8e18bc-558a-4fbe-b909-c9f1f06f91ce)

Reply:
I'm really torn about this. On one hand, React *does* provide a useful error message. On the other hand, it's only on console, and only the first time. But if you're using React components, you should be more willing to look at the console for errors. But, this is BitBurner, and not everyone is familiar with general webdev. *But*, use of React components *is* an advanced concept. Etc...

If there's any way to catch results like this and error-check the input for better messaging *after*, instead of doing a recursive walk of *every* component, that would make me a lot happier.

Reply:
Oh, the one other thing: How do we know that functions are the only thing that errors like this? I'm guessing there are a number of other possible failure modes that act in a similar way. (But maybe they aren't as likely to be hit, so don't justify special handling?)

Reply:
>If there's any way to catch results like this and error-check the input for better messaging after, instead of doing a recursive walk of every component, that would make me a lot happier.

This would really only need to be checked on the initial creation of the react element, not on every rerender cycle.  May need to slightly rework it to function that way, I'm not solid enough on how class components work to know 100% if the constructor gets called every render cycle or just on initial render (I'm pretty sure it's called just once but I might be wrong)

> Oh, the one other thing: How do we know that functions are the only thing that errors like this? I'm guessing there are a number of other possible failure modes that act in a similar way. (But maybe they aren't as likely to be hit, so don't justify special handling?)

I sent every normal type of input in, and many of them did not display anything (boolean t/f, bigint, symbol, undefined, null) but didn't cause any errors. Object types that were not components caused an error, but the error was properly caught by the boundary and displayed normally. Functions were the only inputs in my testing that errored out but not in a way that was caught by the boundary and displayed to the user.

Reply:
> This would really only need to be checked on the initial creation of the react element, not on every rerender cycle.

But for function components, you *are* recreating the element every time the content changes. For instance, if you're using `printRaw` to add some extra styling to your logs, this checking will happen on every print. Potentially multiple times, if you construct your logs with multiple print statements. That's what I'm concerned about.

> I sent every normal type of input in, and many of them did not display anything (boolean t/f, bigint, symbol, undefined, null) but didn't cause any errors.
...
Functions were the only inputs in my testing that errored out but not in a way that was caught by the boundary and displayed to the user.

That's because the functions aren't actually erroring out. They're printing a *warning* to the console (despite the fact that React prints all its warnings with `console.error`) but otherwise ignoring it just like with other bad input such as bigint and symbols.

You could argue that this is poor design on React's part (I would), but I don't think the fact that React shows *extra* debugging info for this one case obligates us to treat it specially. If we are going to do hand-holding-type error checking (which increasingly I feel we shouldn't), it should focus on the bad inputs that React provides *no* feedback for.

Reply:
Followup because I'm concerned I didn't state myself clearly:

I'm all for the new CustomBoundary component, I think that's great! I'm just distinguishing between errors that are *thrown* (and thus potentially crash the game) vs. errors (or in many cases, warnings) that are only printed to the console.

For the latter, there's no way we can detect them after-the-fact, but also no real need. They don't cause *direct* problems from a gameplay perspective; the only issue is that the code isn't doing what the user wanted, and they have to debug that. IMO React has *many* challenges in that department, and we won't be able to find and catch all of them ahead-of-time without requiring the user to watch the console.

Reply:
Yeah you're right about it just being a warning. Objects (that aren't functions) produce an actual error with very similar wording which is part of what was tripping me up there (plus console.error being used for a warning).

![image](https://github.com/bitburner-official/bitburner-src/assets/84951833/7f9c93d1-c099-4c86-bf1a-d8f37ab563f6)

I agree that there's no need to scour the react content to check for warnings, that whole section will be removed.